### PR TITLE
feat (api, worker, hatchery) : let provisioned workers work

### PR DESCRIPTION
**Pull request by yesnault**

### CODE REVIEW

File: engine/api/worker/model.go [hunk 1]
```diff
@@ -80,18 +80,26 @@ func LoadWorkerModelByID(db gorp.SqlExecutor, ID int64) (*sdk.Model, error) {
 }
 
 // LoadWorkerModelsByUser returns worker models list according to user's groups
-func LoadWorkerModelsByUser(db gorp.SqlExecutor, userID int64) ([]sdk.Model, error) {
+func LoadWorkerModelsByUser(db gorp.SqlExecutor, userID int64, isAdmin bool) ([]sdk.Model, error) {
```
Comment (sguiheux) [inline engine/api/worker/model.go]: *sdk.User instead of  userID + isAdmin  ?

File: sdk/hatchery/hatchery.go [hunk 2]
```diff
@@ -73,7 +73,20 @@ func routine(h Interface, provision int, hostname string, timestamp int64, lastS
...
 			defer logTime(fmt.Sprintf("routine> job %d>", job.ID), time.Now(), warningSeconds, criticalSeconds)
-			log.Debug("routine> %d work on job %d", timestamp, job.ID)
+
+			if sdk.IsInArray(job.ID, lastSpawnedIDs) {
+				log.Debug("routine> job %d already spawned in previous routine", job.ID)
+				wg.Done()
+				return
+			}
+
+			if job.QueuedSeconds < 4 {
```
Comment (sguiheux) [inline sdk/hatchery/hatchery.go]: flag

File: engine/api/pipeline/gorp_model.go [hunk 3]
```diff
@@ -35,7 +36,7 @@ func (p *PipelineBuildJob) PostInsert(s gorp.SqlExecutor) error {
 }
 
 //PostGet is a DB Hook on PipelineBuildJob to get jobs and params as from JSON in DB
-func (p *PipelineBuildJob) PostGet(s gorp.SqlExecutor) error {
+func (p *PipelineBuildJob) PostGet(s gorp.SqlExecutor, now time.Time) error {
```
Comment (fsamin) [inline engine/api/pipeline/gorp_model.go]: ko

File: engine/api/worker/model.go [hunk 4]
```diff
@@ -80,18 +80,26 @@ func LoadWorkerModelByID(db gorp.SqlExecutor, ID int64) (*sdk.Model, error) {
 }
 
 // LoadWorkerModelsByUser returns worker models list according to user's groups
-func LoadWorkerModelsByUser(db gorp.SqlExecutor, userID int64) ([]sdk.Model, error) {
+func LoadWorkerModelsByUser(db gorp.SqlExecutor, user sdk.User) ([]sdk.Model, error) {
```
Comment (fsamin) [inline engine/api/worker/model.go]: pointer. and consider if user is nil that it's an admin

File: engine/hatchery/marathon/marathon.go [hunk 5]
```diff
@@ -209,6 +209,7 @@ func (m *HatcheryMarathon) spawnMarathonDockerWorker(model *sdk.Model, hatcheryI
 			"CDS_MODEL":         fmt.Sprintf("%d", model.ID),
 			"CDS_HATCHERY":      fmt.Sprintf("%d", hatcheryID),
 			"CDS_BOOKED_JOB_ID": fmt.Sprintf("%d", job.ID),
+			"CDS_EXIT_ON_END":   "false",
```
Comment (fsamin) [inline engine/hatchery/marathon/marathon.go]: remove

**Pull request merged**